src/program/latex/mod.rs

```diff
@@ -0,0 +1,173 @@
+        }
+    }
+}
+
+pub enum TikzOperator {
+    TikzLeftKet(u32),
+    TikzControl(i32),
+    TikzCnotTarget,
+    TikzCphaseTarget,
+    TikzSwap(i32),
+    TikzSwapTarget,
+    TikzNop,
+    TikzMeasure,
+}
+
+impl TikzOperator {
+    fn get_tikz_operator(tikz_operator: Self) -> String {
+        match tikz_operator {
+            Self::TikzLeftKet(qubit) => format(format_args!(r#"\lstick{{\ket{{q_{{{qubit}}}}}}}"#)), // \lstick{\ket{q_{qubit}}}
+            Self::TikzControl(offset) => format(format_args!(r#"\ctrl{{{offset}}}"#)), // \ctrl{offset}
+            Self::TikzCnotTarget => r"\targ{}".to_string(), // \targ{}
+            Self::TikzCphaseTarget => r"\control{}".to_string(), // \control{}
+            Self::TikzSwap(offset) => format(format_args!(r"\swap{{{offset}}}")), // \swap{offset}
+            Self::TikzSwapTarget => r"\targX{}".to_string(), // \targX{}
+            Self::TikzNop => r"\qw".to_string(),              // \qw
+            Self::TikzMeasure => r"\meter{}".to_string(),     // \meter{}
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum LatexGenError {
+    // TODO: Add variants for each error type using `thiserror` crate to return detailed Result::Err.
+    #[error("This is an error on {qubit_index}.")]
+    SomeError{qubit_index: u32},
+}
+
+pub trait ToLatex {
+    fn to_latex(self, diagram_settings: DiagramSettings) -> Result<String, LatexGenError>;
+}
+
+impl ToLatex for Program {
+    fn to_latex(self, diagram_settings: DiagramSettings) -> Result<String, LatexGenError> {
+        // TODO: Generate the Program LaTeX.
+        let latex = "";
+
+        Ok(latex.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{DiagramSettings, ToLatex};
+    use crate::Program;
+    use std::str::FromStr;
+
+    /// Take an instruction and return the LaTeX using the to_latex method.
+    pub fn get_latex(instructions: &str) -> String {
+        let program = Program::from_str(instructions).expect("Program should be returned.");
+        program
+            .to_latex(DiagramSettings::default())
+            .expect("LaTeX should generate without error.")
+    }
+
+    #[test]
+    /// Test functionality of to_latex using default settings.
+    fn test_to_latex() {
+        let program = Program::from_str("").expect("");
+        program.to_latex(DiagramSettings::default()).expect("");
+    }
+
+    mod gates {
+        use crate::program::latex::tests::get_latex;
+
+        #[test]
+        fn test_gate_x() {
+            insta::assert_snapshot!(get_latex("X 0"));
+        }
+
+        #[test]
+        fn test_gate_y() {
+            insta::assert_snapshot!(get_latex("Y 1"));
+        }
+
+        #[test]
+        fn test_gate_controlled() {
+            insta::assert_snapshot!(get_latex("CONTROLLED H 3 2"));
+        }
+    }
+
+    mod tikz_operators {
+        use crate::program::latex::TikzOperator;
+
+        #[test]
+        fn test_tikz_left_ket() {
```

Comment:
Worth noting before you write too many more of these - macros are quite useful for test case patterns like this. There is a learning curve but it's a powerful tool in the kit

Reply:
Thanks for watching out Kalan! I haven't written any more tests since I started the new design. I'll look into macros once I get around to writing more.